app/models/learning.rb

```diff
@@ -1,12 +1,24 @@
 # frozen_string_literal: true
 
 class Learning < ActiveRecord::Base
-  enum status: { started: 0, complete: 1, not_complete: 2 }
+  enum status: { not_complete: 0, started: 1, submitted: 2, complete: 3 }
   belongs_to :user, touch: true
   belongs_to :practice
 
   validates :user_id, presence: true
   validates :practice_id,
     presence: true,
     uniqueness: { scope: :user_id }
+  validate :startable_practice
+
+  def change_status(status)
+    self.update(status: status)
+  end
+
+  private
+    def startable_practice
+      unless Learning.find_by(user_id: self.user_id, status: "started").nil? || self.status != "started"
```

Comment:
```suggestion
      unless !Learning.where(user_id: self.user_id, status: "started").exists? || self.status != "started"
```

https://railsguides.jp/active_record_querying.html#%E3%82%AA%E3%83%96%E3%82%B8%E3%82%A7%E3%82%AF%E3%83%88%E3%81%AE%E5%AD%98%E5%9C%A8%E3%83%81%E3%82%A7%E3%83%83%E3%82%AF

にすると無駄にLearningのインスタンスを作らずに済み効率的です。
ログでSQLの違いも確認してみてください。


```ruby
unless Learning.where(user_id: self.user_id, status: "started").count.zero? || self.status != "started"
```

と同じです。こっちの方が読みやすいならこっちでも良いです

Reply:
[enum](https://api.rubyonrails.org/v5.2.3/classes/ActiveRecord/Enum.html)は色んな機能をサポートしているので、シンプルに書けるかもです！

```
# Learning.find_by(user_id: self.user_id, status: "started")
Learning.started.find_by(user_id: user_id)

# self.status != "started"
!started?

# おまけ（idの省略）
Learning.find_by(user: user)
```

あと、ちょっと別件になるかもですが、Rails 6からenumに[negative scopesがサポート](https://github.com/rails/rails/pull/35381)されたので、`not_complete` という名称だと誤解を招く可能性があるかもです…！

```
# `complete` で抽出
Learning.complete

# `not_complete`で抽出（`complete`以外を抽出が動かない）
Learning.not_complete

# `not_complete`以外を抽出
Learning.not_not_complete
```

Reply:
@bluerabbit さん
@rono23 さん
ご指摘ありがとうございます🙇
頂いたご指摘をもとに修正しました。

存在確認だけであればご指摘いただいたほうが効率的ということがわかりました。
`SELECT  1 AS one FROM "learnings" WHERE "learnings"."status" = $1 AND "learnings"."user_id" = $2 LIMIT $3  [["status", 1], ["user_id", 459775584], ["LIMIT", 1]]`
SQLも`select 1`になるんですね。

enumに関しても知らない書き方や使い方があったのでとても勉強になりました。
`not_complete`の名称も併せて検討します…

Reply:
> enumは色んな機能をサポートしているので、シンプルに書けるかもです！

Wow, cool!

> negative scopes

なんと・・・

Reply:
> Learning.started.find_by(user_id: user_id)

こんな書き方できたんですね 📝